src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs

```diff
@@ -1601,40 +1601,46 @@ private static bool IsAzureDevOpsArtifactsUrl(string location)
                     Directory.CreateDirectory(directory);
                 }
 
-                if (authHeader != null)
-                {
-                    requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri)
-                    {
-                        Headers =
-                        {
-                            { HttpRequestHeader.Authorization.ToString(), authHeader.ToString() }
-                        }
-                    };
-                }
-                else
-                {
-                    requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri);
-                }
-
                 // Ensure the parent target directory has been created.
                 using (FileStream outStream = new FileStream(temporaryFileName,
-                                                      _options.Overwrite ? FileMode.Create : FileMode.CreateNew,
-                                                      FileAccess.Write))
+                                                        _options.Overwrite ? FileMode.Create : FileMode.CreateNew,
+                                                        FileAccess.Write))
                 {
-                    using (var response = await client.SendAsync(requestMessage))
-                    {
-                        response.EnsureSuccessStatusCode();
-
-                        using (var inStream = await response.Content.ReadAsStreamAsync())
+                    await ExponentialRetry.Default.RetryAsync(
+                        async () =>
                         {
-                            downloadOutput.AppendLine($"    {sourceUri} =>");
-                            foreach (string targetFile in targetFiles)
+                            if (authHeader != null)
                             {
-                                downloadOutput.AppendLine($"      {targetFile}");
+                                requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri)
+                                {
+                                    Headers =
+                                    {
+                                        { HttpRequestHeader.Authorization.ToString(), authHeader.ToString() }
+                                    }
+                                };
                             }
-                            await inStream.CopyToAsync(outStream, cancellationToken);
-                        }
-                    }
+                            else
+                            {
+                                requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri);
+                            }
+
+                            using (var response = await client.SendAsync(requestMessage))
+                            {
+                                response.EnsureSuccessStatusCode();
+
+                                using (var inStream = await response.Content.ReadAsStreamAsync())
```

Comment:
It feels like this chunk should be outside the retry.  The retry should only contain the "web" stuff.  Then we'd get the resulting object as a result of the retry and use the returned data somehow? I don't know what ExponentialRetry does exactly, but it feels like the only stuff inside the "retry this" scope is the stuff we are worried might fail.

It would be weird if we wrote half the stuff to the file, then something happened and we retried at that point, which would then have one and a half copies of the output copied to it, likely leading to very confusing crashes later on, or worse, wrong behavior.

Reply:
Yeah I think so. If it's inside, then we probably need to pull the deletion inside there too.

Reply:
I put it inside because otherwise I was getting these weird exceptions that said:

```
Error while copying content to a stream.
```

I believe it was because I was returning the response, which led me to moving the stream inside the retry. Of course, that's also an HttpRequestException, so it would retry it, but that felt like unnecessary retries.

Reply:
Yeah, since the inner code disposes the response object, you can't use it outside.  You'd have to return something else (or not dispose the response inside in "success" cases)

Reply:
Oh, I bet the response and associated streams were getting disposed while exiting the retry block.

Reply:
No, I was doing that. It turns out that even this way, I'm getting the same errors. The original code looked like this:

```
using (var response = await ExponentialRetry.Default.RetryAsync(
                        async () =>
                        {
                            if (authHeader != null)
                            {
                                requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri)
                                {
                                    Headers =
                                    {
                                        { HttpRequestHeader.Authorization.ToString(), authHeader.ToString() }
                                    }
                                };
                            }
                            else
                            {
                                requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri);
                            }

                            var response = await client.SendAsync(requestMessage);
                            response.EnsureSuccessStatusCode();

                            if (requestMessage != null)
                            {
                                requestMessage.Dispose();
                            }

                            return response;
                        },
                        ex => Console.WriteLine($"    Failed to download {sourceUri}: {ex.Message}"),
                        ex => ex is HttpRequestException))
                    {
                        using (var inStream = await response.Content.ReadAsStreamAsync())
                        {
                            downloadOutput.AppendLine($"    {sourceUri} =>");
                            foreach (string targetFile in targetFiles)
                            {
                                downloadOutput.AppendLine($"      {targetFile}");
                            }
                            await inStream.CopyToAsync(outStream, cancellationToken);
                        }
                    }
```

Reply:
It doesn't happen all the time, and is successful on retry. So it was kind of confusing to see.

Reply:
That original code doesn't appear to have _any_ dispose call for the response in it, so it's a little weird that you'd have problems reading back the value.

Reply:
@ChadNedzlek It's possible that when the retry was added, the disposal was happening the background. It's not guaranteed, but it could happen.

Reply:
Background disposal wouldn't have happend if you still had a GC handle to the object.  Since the HttpClient and HttpResponse were still bound to in scope variables, that shouldn't have been happening.  I'd have to know what the actual exception was to be much more use. :-) "Error copying the stream" isn't much.

Reply:
I'm going to see if there's an innerException/anything interesting in there. Maybe we can get more info.

Reply:
Stack Trace:
```
at System.Net.Http.HttpContent.LoadIntoBufferAsyncCore(Task serializeToStreamTask, MemoryStream tempBuffer)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.DotNet.Darc.Operations.GatherDropOperation.<>c__DisplayClass49_0.<<DownloadFileImplAsync>b__0>d.MoveNext() in D:\michelm\arcade-services\src\Microsoft.DotNet.Darc\src\Darc\Operations\GatherDropOperation.cs:line 1618
```

So basically, during the SendAsync

Reply:
"Forcibly closed by the remote host" means the other end of the connection said to go away.  That can happen if you take too long to read the data, and the server decided you were trolling it, and killed your connection.  This code isn't weirdly slow or anything, is it?

Reply:
It can be. but also, i keep running it over and over again, so maybe it's just mad :)